### PR TITLE
[jdk_17_migration] Migrate Ballerina to Java 17

### LINKED ISSUE #40387: [Improvement]: Migrate to Java 17
### Description

We need to migrate to Java 17 from Java 11. Need to test compiler and runtime and other related tools.

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #40387

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-cli/src/test/resources/test-resources/print-utils-output/search-results.json [hunk 1]
```diff
@@ -41,7 +41,7 @@
...
...\nSF_BASE_URL=\"<Salesforce base url (eg: https:\/\/ap15.salesforce.com)>\"\nSF_ACCESS_TOKEN=\"<Salesforce access token>\"\nSF_CLIENT_ID=\"<Salesforce client ID>\"\nSF_CLIENT_SECRET=\"<Salesforce client secret>\"\nSF_REFRESH_URL=\"<Salesforce refresh url (eg: https:\/\/login.salesforce.com\/services\/oauth2\/token)>\"\nSF_REFRESH_TOKEN=\"<Salesforce refresh token>\"\nJDBC_URL=\"<JDBC URL (eg: jdbc:mysql:\/\/localhost:3306\/sf_company)>\"\nDB_USERNAME=\"<MySQL database username>\"\nDB_PASSWORD=\"<MySQL database password>\"\nSCHEDULER_INTERVAL_IN_MILLIS=<Scheduler interval in milli-seconds (eg: 60000)>\n```\n\nLet’s build the module. Navigate to the project root directory and execute the following command.\n\n```bash\n$ bal build salesforce_to_mysql\n```\n\nThis creates the executables. Now run the `salesforce_to_mysql.jar` file created in the above step.\n\n```bash\n$ java -jar target\/bin\/salesforce_to_mysql.jar\n```\n\nYou will see the following log after successfully updating the database.\n\n```\n2019-09-26 17:41:27,708 INFO  [wso2\/sfdc_to_mysql_db] - service started...\n2019-09-26 17:41:32,094 INFO  [wso2\/sfdc_to_mysql_db] - Batch job SFDC -> MySQL has been completed.\n```",
```
Comment (warunalakshitha) [inline cli/ballerina-cli/src/test/resources/test-resources/print-utils-output/search-results.json:44]: Why this change?

File: distribution/zip/jballerina-tools/build.gradle [hunk 2]
```diff
@@ -507,17 +507,17 @@ artifacts {
...
-    configurations.dist.dependencies
-        .findAll { it instanceof ProjectDependency }
-        .each {
-            it.dependencyProject.afterEvaluate {
-                def dependencyCheck = it.tasks.find { it.name == 'check' }
-                check.dependsOn dependencyCheck
-            }
-        }
-}
+////// This section makes sure that checkstyle/spotbug runs when dist is build.
```
Comment (warunalakshitha) [inline distribution/zip/jballerina-tools/build.gradle:510]: Can we remove this?

File: distribution/zip/jballerina/build.gradle [hunk 3]
```diff
@@ -156,13 +156,13 @@ build {
...
-    configurations.dist.dependencies
-        .findAll { it instanceof ProjectDependency }
-        .each {
-            it.dependencyProject.afterEvaluate {
-                def dependencyCheck = it.tasks.find { it.name == 'check' }
-                check.dependsOn dependencyCheck
-            }
-        }
-}
+//project.afterEvaluate {
```
Comment (warunalakshitha) [inline distribution/zip/jballerina/build.gradle:159]: Any reason to comment out this?

File: distribution/zip/jballerina-tools/build.gradle (hunk 2 above)
Comment (gabilang) [reply distribution/zip/jballerina-tools/build.gradle:510]: I'll revisit the commented out lines and will remove if they are not necessary

File: misc/diagram-util/build.gradle [hunk 4]
```diff
@@ -21,9 +21,9 @@ configurations {
     dist {
         transitive false
     }
-    dependency {
-        transitive false
-    }
+//    dependency {
```
Comment (warunalakshitha) [inline misc/diagram-util/build.gradle:24]: Shall we remove this?

File: misc/diagram-util/build.gradle [hunk 5]
```diff
@@ -64,3 +54,8 @@ test {
         suites 'src/test/resources/testng.xml'
     }
 }
+
+///*this is to let gradle infer the locations of the modules*/
+//java {
```
Comment (warunalakshitha) [inline misc/diagram-util/build.gradle:59]: Can we remove this?

File: misc/docerina/build.gradle [hunk 6]
```diff
@@ -16,7 +16,7 @@
  */
  
 apply plugin: 'base'
-apply plugin: 'com.github.johnrengelman.shadow'
+//apply plugin: 'com.github.johnrengelman.shadow'
```
Comment (warunalakshitha) [inline misc/docerina/build.gradle:19]: Why we do not need this plugin?

File: misc/diagram-util/build.gradle (hunk 5 above)
Comment (gabilang) [reply misc/diagram-util/build.gradle:59]: Earlier also there are many lines has been commented. I'll remove them in later commits

File: misc/testerina/modules/testerina-runtime/build.gradle [hunk 7]
```diff
@@ -53,13 +54,3 @@ dependencies {
...
-
-compileJava {
-    inputs.property("moduleName", moduleName)
-    doFirst {
-        options.compilerArgs = [
-                '--module-path', classpath.asPath,
-        ]
-        classpath = files()
-    }
-}
```
Comment (warunalakshitha) [inline misc/testerina/modules/testerina-runtime/build.gradle:65]: Lets add new line here

File: misc/docerina/build.gradle (hunk 6 above)
Comment (gabilang) [reply misc/docerina/build.gradle:19]: We removed the shadowJar task from there

File: project-api/project-api-test/src/test/java/io/ballerina/projects/test/plugins/CompilerPluginTests.java [hunk 8]
```diff
@@ -100,7 +100,7 @@ public void init() {
     }
 
     @Test
-    public void testCompilerPluginBasic() {
```
Comment (warunalakshitha) [inline project-api/project-api-test/src/test/java/io/ballerina/projects/test/plugins/CompilerPluginTests.java:103]: Do we need this change?

File: project-api/project-api-test/src/test/resources/balaloader/foo-winery-any-0.1.0.bala
Comment (warunalakshitha) [inline project-api/project-api-test/src/test/resources/balaloader/foo-winery-any-0.1.0.bala:1]: We can remove this balo file

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/TypedescriptorTest.java [hunk 9]
```diff
@@ -812,14 +812,14 @@ private Object[][] getDistinctObjectTypes() {
...
-    public void testCompileErrorType1() {
+    public void testImplementationErrorType1() {
         LineRange range = LineRange.from("typedesc_test.bal", from(181, 12), from(181, 17));
         Optional<TypeSymbol> type = model.type(range);
         assertEquals(type.get().typeKind(), COMPILATION_ERROR);
     }
 
     @Test
-    public void testCompileErrorType2() {
+    public void testImplementationErrorType2() {
```
Comment (warunalakshitha) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/TypedescriptorTest.java:822]: These changes are not required

File: tests/jballerina-integration-test/src/test/resources/async/asyncFunctionPackage/Dependencies.toml [hunk 10]
```diff
@@ -5,6 +5,7 @@
 
 [ballerina]
 dependencies-toml-version = "2"
+distribution-version = "2201.6.0-SNAPSHOT"
```
Comment (warunalakshitha) [inline tests/jballerina-integration-test/src/test/resources/async/asyncFunctionPackage/Dependencies.toml:8]: Do we need these changes?

File: tests/jballerina-unit-test/src/test/resources/test-src/object/PrivateFieldsAccessProject/object-private-fields-02-negative.bal [hunk 11]
```diff
@@ -22,7 +22,7 @@ public class personFoo {
     }
 }
 
-public function testCompileTimeStructEq() {
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/PrivateFieldsAccessProject/object-private-fields-02-negative.bal:25]: We can revert this change other places as well

File: misc/testerina/modules/testerina-runtime/build.gradle (hunk 7 above)
Comment (gabilang) [reply misc/testerina/modules/testerina-runtime/build.gradle:65]: Here actually I removed the trailing new line not found error by adding a new line at the end of this file.

File: cli/ballerina-cli/src/test/resources/test-resources/print-utils-output/search-results.json (hunk 1 above)
Comment (gabilang) [reply cli/ballerina-cli/src/test/resources/test-resources/print-utils-output/search-results.json:44]: It's due to target JVM change to java17

File: distribution/zip/jballerina-tools/build.gradle (hunk 2 above)
Comment (gabilang) [reply distribution/zip/jballerina-tools/build.gradle:510]: Evaluating a task second time has been removed from the Gradle version 7 and we need to make sure spotbugs and checkstyle checks run on the dist build.
